digits.t-mobile.com - Safari is not a recommended browser

colelao opened the pull request:
#### a579d4b73ccf9843bc5efb8ac1974dd05c465f28
<pre>
digits.t-mobile.com - Safari is not a recommended browser
<a href="https://bugs.webkit.org/show_bug.cgi?id=276940">https://bugs.webkit.org/show_bug.cgi?id=276940</a>
<a href="https://rdar.apple.com/132309517">rdar://132309517</a>

Reviewed by Brent Fulgham.

This enables the Navigator UserAgentData API only when the needsNavigatorUserAgentDataQuirk is enabled.
This quirk is enabled on digits.t-mobile.com is loaded.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::isStorageAccessQuirkDomainAndElement):
(WebCore::Quirks::needsNavigatorUserAgentDataQuirk const):
(WebCore::handleTMobileQuirks):
(WebCore::handleGoogleQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/page/UserAgentStringParser.cpp:
(WebCore::UserAgentStringParser::populateUserAgentData):

Canonical link: <a href="https://commits.webkit.org/298941@main">https://commits.webkit.org/298941@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4e91ac4d8e4b458a051d3c3cb1f9449222272559

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123326 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119103 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37594 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45485 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88967 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120161 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105115 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28983 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23228 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67000 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109333 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99327 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23403 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126448 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115735 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44125 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33149 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97638 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97430 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24813 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42775 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20707 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43998 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49657 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144435 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43454 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37176 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46799 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45150 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5301,9 +5301,10 @@ NavigationAPIEnabled:
 NavigatorUserAgentDataJavaScriptAPIEnabled:
   type: bool
-  status: unstable
+  status: testable
```
This should be `internal`, since we don't really want a toggle to show up for people to activate; it's just for Quirk use.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -999,17 +999,17 @@ static bool isStorageAccessQuirkDomainAndElement(const URL& url, const Element&
-        || element.hasClassName("mectrl_header"_s));
+            && (element.hasClassName("glyph_signIn_circle"_s)
+                || element.hasClassName("mectrl_headertext"_s)
+                || element.hasClassName("mectrl_header"_s));
```
Nit: You shouldn't include these unrelated whitespace changes.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -999,17 +999,17 @@ static bool isStorageAccessQuirkDomainAndElement(const URL& url, const Element&
-        || element.hasClassName("sb-signin-button"_s));
+            && (element.hasClassName("web-toolbar__signin-button"_s)
+                || element.hasClassName("web-toolbar__signin-button-label"_s)
+                || element.hasClassName("sb-signin-button"_s));
```
Ditto. Might need to check your editor settings.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -3041,7 +3064,7 @@ void Quirks::determineRelevantQuirks()
         { "ralphlauren"_s, &handleRalphLaurenQuirks },
 #endif
 #if ENABLE(VIDEO_PRESENTATION_MODE)
-        { "reddit"_s, & handleRedditQuirks },
```
I do approve of this unrelated whitespace change, though :-)

brentfulgham reviewed (approved):
This is great! Please switch the feature flag to `internal`, but otherwise this is ready to go! r=me

colelao committed:
digits.t-mobile.com - Safari is not a recommended browser
https://bugs.webkit.org/show_bug.cgi?id=276940
rdar://132309517

Reviewed by Brent Fulgham.

This enables the Navigator UserAgentData API only when the needsNavigatorUserAgentDataQuirk is enabled.
This quirk is enabled on digits.t-mobile.com is loaded.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/page/Navigator.cpp:
(WebCore::Navigator::userAgentData const):
* Source/WebCore/page/Navigator.idl:
* Source/WebCore/page/NavigatorUAData.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::isStorageAccessQuirkDomainAndElement):
(WebCore::Quirks::needsNavigatorUserAgentDataQuirk const):
(WebCore::handleTMobileQuirks):
(WebCore::handleGoogleQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:
* Source/WebCore/page/UserAgentStringParser.cpp:
(WebCore::UserAgentStringParser::populateUserAgentData):

Canonical link: https://commits.webkit.org/298941@main
